### ISSUE TITLE
Empty ground truth problem

### ISSUE THREAD

**Issue by JinyuCai124578** [janelia-cellmap/cellmap-segmentation-challenge#148]
Is it right that the ground truth is (almost) empty in most er and nuc labels?

![Image](https://github.com/user-attachments/assets/6cfec9f6-ff4b-4465-b09f-eba3a335b10f)

![Image](https://github.com/user-attachments/assets/8814a615-7860-4e32-83b6-d141ab092c8f)

By almost I mean some do exist a ground truth, but it only occupies a tiny area and doesn't seem right.

![Image](https://github.com/user-attachments/assets/860d5b14-3c7f-485a-8e46-f66fb5144d45)

And some of the raw images are misty white. Is it something to do with the data quality?

![Image](https://github.com/user-attachments/assets/b30c19b4-4e50-4f3a-95e2-313dc7395ede)

**Comment by rhoadesScholar** [comment]
The misty white has to do with the normalization of the image relative to the 0 background I believe - and the fact that some portions have very little contrast, such as the interior of the nucleus or extracellular space.

@JinyuCai124578 As for the (almost) emptiness, can you please supply a minimal script to find examples so I can better diagnose this?

**Comment by rhoadesScholar** [comment]
For instance, running the following snippet:
```
# %%
import os
from upath import UPath as Path
import cellmap_segmentation_challenge as csc
from cellmap_data.transforms.augment import (
    RandomContrast,
    GaussianNoise,
    Normalize,
    NaNtoNum,
    GaussianBlur,
    Binarize,
)
import torchvision.transforms.v2 as T
import torch
import numpy as np
import matplotlib.pyplot as plt

force_remake_datasplit = True

classes = [
    # "ecs",
    # "pm",
    # "cyto",
    # "golgi",
    # "ves",
    # "endo",
    # "lyso",
    # "ld",
    # "mito",
    "nuc",
    "er",
]  # list of classes to segment

classes.sort()
# force_all_classes = "validate"
force_all_classes = True
pad = True
seed = 13

validation_prob = 0.15
only_named_classes = True

# %%
# Define the spatial transforms (augmentations) to be applied to the data for training

# Define the spatial transformations to apply to the training data
# spatial_transforms = {
#     "mirror": {"axes": {"x": 0.5, "y": 0.5, "z": 0.5}},
#     "transpose": {"axes": ["x", "y", "z"]},
#     "rotate": {"axes": {"x": [-180, 180], "y": [-180, 180], "z": [-180, 180]}},
# }
spatial_transforms = {
    "mirror": {"axes": {"x": 0.5, "y": 0.5}},
    "transpose": {"axes": ["x", "y"]},
    "rotate": {"axes": {"x": [-180, 180], "y": [-180, 180]}},
}
train_raw_value_transforms = T.Compose(
    [
        T.ToDtype(torch.float),
        Normalize(),
        NaNtoNum({"nan": 0, "posinf": None, "neginf": None}),
        T.RandomApply(
            [
                RandomContrast(),
                GaussianNoise(std=0.04),
            ]
        ),
    ],
)
target_value_transforms = T.Compose(
    [
        T.ToDtype(torch.float),
        Binarize(threshold=0.5),
        GaussianBlur(kernel_size=11, sigma=5, dim=3),
    ],
)

class_relation_dict = csc.utils.get_class_relations(
    named_classes=classes if only_named_classes else None
)


# Define the paths to the raw and groundtruth data and the label classes by crawling the directories and writing the paths to a csv file
csv_path = os.path.join(Path(__file__).parent, "datasplit.csv")
if not os.path.exists(csv_path):
    csc.utils.make_datasplit_csv(
        validation_prob=validation_prob,
        classes=classes,
        csv_path=csv_path,
        force_all_classes=force_all_classes,
    )
elif not force_remake_datasplit:
    print(f"Using existing datasplit csv file: {csv_path}")
else:
    print(f"Remaking datasplit csv file: {csv_path}")
    # Remove the existing csv file and create a new one
    os.remove(csv_path)
    csc.utils.make_datasplit_csv(
        validation_prob=validation_prob,
        classes=classes,
        csv_path=csv_path,
        force_all_classes=force_all_classes,
    )

# %%

input_key = "input"  # key of the input data
output_key = "output"  # key of the target data
scale = (8, 8, 8)  # voxel size of the data
# scale = (16, 16, 16)  # voxel size of the data
input_array_info = {
    "shape": (128, 128, 128),
    "scale": scale,
}  # shape and voxel size of the data to load for the input
target_array_info = {
    "shape": (128, 128, 128),
    "scale": scale,
}  # shape and voxel size of the data to load for the target
# %%
from cellmap_data.utils.figs import get_image_grid


batch_size = 4
train_loader, val_loader = csc.utils.get_dataloader(
    csv_path,
    classes,
    batch_size,
    input_array_info=input_array_info,
    target_array_info=target_array_info,
    spatial_transforms=spatial_transforms,
    target_value_transforms=target_value_transforms,
    train_raw_value_transforms=train_raw_value_transforms,
    use_mutual_exclusion=True,
)


# %%
train_loader.refresh()
d = np.random.randint(len(train_loader.dataset.datasets))
print(d)
print(train_loader.dataset.datasets[d])

# %%
# Visualize an example of the training data
test = next(iter(train_loader.loader))


def good_batch(batch):
    batch[torch.isnan(batch.cpu())] = 0
    return batch.sum() >= np.prod(batch.shape) * 0.1


count = 1
while not good_batch(test[output_key].clone()):
    test = next(iter(train_loader.loader))
    count += 1

print(f"Found a good batch after {count} iterations")
_ = get_image_grid(
    test[input_key],  # [None, ...],
    test[output_key],  # [None, ...],
    test[output_key],  # [None, ...],
    classes,
    batch_size=batch_size,
    clim=[0, 1],
)

```
I get this:

<img width="584" alt="Image" src="https://github.com/user-attachments/assets/f63df2c3-3917-4212-a143-f15f3d7f4263" />

**Comment by rhoadesScholar** [comment]
@JinyuCai124578 Is this still an ongoing issue?

**Comment by JinyuCai124578** [comment]
(I've fixed the upper problem by upgrade cellmap_data.)
I am trying to calculate the proportion of bad batches with your snippet,
`"""
This script is designed for preparing and validating a dataset for a cell segmentation challenge. 
It includes the following functionalities:

1. Imports necessary libraries and modules for data processing and augmentation.
2. Defines parameters for data splitting, class selection, and spatial transformations.
3. Creates or refreshes a CSV file containing paths to the raw and ground truth data.
4. Sets up data loaders for training and validation datasets with specified transformations.
5. Visualizes a sample batch of training data to ensure the data is suitable for training.

Key components:
- Data augmentation techniques such as normalization, random contrast, and Gaussian noise.
- Class relations management for segmentation tasks.
- Data loading and preprocessing to handle input and target data efficiently.
"""
# %%
import os
from upath import UPath as Path
import cellmap_segmentation_challenge as csc
from cellmap_data.transforms.augment import (
    RandomContrast,
    GaussianNoise,
    Normalize,
    NaNtoNum,
    GaussianBlur,
    Binarize,
)
import torchvision.transforms.v2 as T
import torch
import numpy as np
import matplotlib.pyplot as plt

# Defines parameters for data splitting, class selection, and spatial transformations.
force_remake_datasplit = True

classes = [
    # "ecs",
    # "pm",
    # "cyto",
    # "golgi",
    # "ves",
    # "endo",
    # "lyso",
    # "ld",
    # "mito",
    "nuc",
    "er",
]  # list of classes to segment

classes.sort()
# force_all_classes = "validate"
force_all_classes = True
pad = True
seed = 13

validation_prob = 0.15
only_named_classes = True

# %%
# Define the spatial transforms (augmentations) to be applied to the data for training

# Define the spatial transformations to apply to the training data
# spatial_transforms = {
#     "mirror": {"axes": {"x": 0.5, "y": 0.5, "z": 0.5}},
#     "transpose": {"axes": ["x", "y", "z"]},
#     "rotate": {"axes": {"x": [-180, 180], "y": [-180, 180], "z": [-180, 180]}},
# }
spatial_transforms = {
    "mirror": {"axes": {"x": 0.5, "y": 0.5}},
    "transpose": {"axes": ["x", "y"]},
    "rotate": {"axes": {"x": [-180, 180], "y": [-180, 180]}},
}
train_raw_value_transforms = T.Compose(
    [
        T.ToDtype(torch.float),
        Normalize(),
        NaNtoNum({"nan": 0, "posinf": None, "neginf": None}),
        T.RandomApply(
            [
                RandomContrast(),
                GaussianNoise(std=0.04),
            ]
        ),
    ],
)
target_value_transforms = T.Compose(
    [
        T.ToDtype(torch.float),
        Binarize(threshold=0.5),
        GaussianBlur(kernel_size=11, sigma=5, dim=3),
    ],
)

class_relation_dict = csc.utils.get_class_relations(
    named_classes=classes if only_named_classes else None
)


# Define the paths to the raw and groundtruth data and the label classes by crawling the directories and writing the paths to a csv file
csv_path = os.path.join(Path(__file__).parent, "datasplit.csv")
if not os.path.exists(csv_path):
    csc.utils.make_datasplit_csv(
        validation_prob=validation_prob,
        classes=classes,
        csv_path=csv_path,
        force_all_classes=force_all_classes,
    )
elif not force_remake_datasplit:
    print(f"Using existing datasplit csv file: {csv_path}")
else:
    print(f"Remaking datasplit csv file: {csv_path}")
    # Remove the existing csv file and create a new one
    os.remove(csv_path)
    csc.utils.make_datasplit_csv(
        validation_prob=validation_prob,
        classes=classes,
        csv_path=csv_path,
        force_all_classes=force_all_classes,
    )

# %%

input_key = "input"  # key of the input data
output_key = "output"  # key of the target data
scale = (8, 8, 8)  # voxel size of the data
# scale = (16, 16, 16)  # voxel size of the data
input_array_info = {
    "shape": (128, 128, 128),
    "scale": scale,
}  # shape and voxel size of the data to load for the input
target_array_info = {
    "shape": (128, 128, 128),
    "scale": scale,
}  # shape and voxel size of the data to load for the target
# %%
from cellmap_data.utils.figs import get_image_grid


batch_size = 4
train_loader, val_loader = csc.utils.get_dataloader(
    csv_path,
    classes,
    batch_size,
    input_array_info=input_array_info,
    target_array_info=target_array_info,
    spatial_transforms=spatial_transforms,
    target_value_transforms=target_value_transforms,
    train_raw_value_transforms=train_raw_value_transforms,
    use_mutual_exclusion=True,
)


# %%
train_loader.refresh()
d = np.random.randint(len(train_loader.dataset.datasets))
print(d)
print(train_loader.dataset.datasets[d])

# %%
# Visualize an example of the training data
loader_iter = iter(train_loader.loader)  # Create an iterator from the loader
test = next(loader_iter)  # Get the next batch from the iterator


def good_batch(batch):
    batch[torch.isnan(batch.cpu())] = 0
    return batch.sum() >= np.prod(batch.shape) * 0.1

from tqdm import tqdm
bad_batch_count = 0
total_batch=0
for test in tqdm(train_loader.loader):
    if good_batch(test[output_key].clone()):
        continue
    else:
        bad_batch_count += 1
        fig = get_image_grid(
            test[input_key],  # [None, ...],
            test[output_key],  # [None, ...],
            test[output_key],  # [None, ...],
            classes,
            batch_size=batch_size,
            clim=[0, 1],
        )

        # Save the figure
        fig.savefig(f"output_figure_bad{bad_batch_count}.png")  # Specify the desired filename and format
    total_batch += 1
    tqdm.write(f"Bad batch encountered: {bad_batch_count} / {total_batch}")
print(f"Total bad batches encountered: {bad_batch_count}")
fig = get_image_grid(
    test[input_key],  # [None, ...],
    test[output_key],  # [None, ...],
    test[output_key],  # [None, ...],
    classes,
    batch_size=batch_size,
    clim=[0, 1],
)

# Save the figure
fig.savefig("output_figure.png")  # Specify the desired filename and format

# bad_batch_count = 0
# count = 1
# while not good_batch(test[output_key].clone()):
#     test = next(iter(train_loader.loader))
#     bad_batch_count += 1
#     count += 1

# print(f"Found a good batch after {count} iterations")
# print(f"Total bad batches encountered: {bad_batch_count}")
# _ = get_image_grid(
#     test[input_key],  # [None, ...],
#     test[output_key],  # [None, ...],
#     test[output_key],  # [None, ...],
#     classes,
#     batch_size=batch_size,
#     clim=[0, 1],
# )`
But

![Image](https://github.com/user-attachments/assets/2db4e6a8-d62f-4feb-8834-1c240259263c)
And most nuc preds are empty except one.

![Image](https://github.com/user-attachments/assets/4520bc47-c692-42e7-9c00-a8820d244763)

Is that right? At least I can't find a good batch in 2 iterations.
